framework/packages/abstract-std/src/native/ibc/ica_client.rs

```diff
@@ -1 +1,170 @@
-pub mod state {}
+use crate::objects::TruncatedChainId;
+use cosmwasm_schema::QueryResponses;
+use cosmwasm_std::CosmosMsg;
+use cw_ownable::{cw_ownable_execute, cw_ownable_query};
+
+pub use action::{IcaAction, IcaActionResponse, IcaExecute};
+pub use chain_type::{CastChainType, ChainType};
+
+pub use polytone_evm::EVM_NOTE_ID;
+pub use polytone_evm::POLYTONE_EVM_VERSION;
+
+/// This needs no info. Owner of the contract is whoever signed the InstantiateMsg.
+#[cosmwasm_schema::cw_serde]
+pub struct InstantiateMsg {}
+
+#[cosmwasm_schema::cw_serde]
+pub enum MigrateMsg {
+    Instantiate(InstantiateMsg),
+    Migrate {},
+}
+
+#[cw_ownable_execute]
+#[cosmwasm_schema::cw_serde]
+#[derive(cw_orch::ExecuteFns)]
+pub enum ExecuteMsg {}
+
+#[cw_ownable_query]
+#[cosmwasm_schema::cw_serde]
+#[derive(QueryResponses, cw_orch::QueryFns)]
+pub enum QueryMsg {
+    /// Returns config
+    /// Returns [`ConfigResponse`]
+    #[returns(ConfigResponse)]
+    Config {},
+
+    #[returns(IcaActionResult)]
+    IcaAction {
+        // Account address address used to query polytone implementations or proxy itself.
+        account_address: String,
+        // Chain to send to
+        chain: TruncatedChainId,
+        // Queries go first
+        actions: Vec<IcaAction>,
+    },
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct ConfigResponse {
+    pub ans_host_address: String,
+    pub registry_address: String,
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct IcaActionResult {
+    /// messages that call the underlying implementations (be it polytone/cw-ica-controller/etc)
+    pub msgs: Vec<CosmosMsg>,
+}
+
+mod chain_type {
+    use std::fmt::Display;
+
+    use crate::constants::*;
+    use crate::objects::TruncatedChainId;
+
+    #[derive(Debug, Clone, Copy, PartialEq)]
+    pub enum ChainType {
+        Evm,
+        Cosmos,
+    }
+
+    impl Display for ChainType {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            match self {
+                ChainType::Evm => write!(f, "EVM"),
+                ChainType::Cosmos => write!(f, "Cosmos"),
+            }
+        }
+    }
+
+    pub trait CastChainType {
+        fn chain_type(&self) -> Option<ChainType>;
+    }
+
+    impl CastChainType for TruncatedChainId {
+        // Return the type of chain based on the chain-id.
+        // Note: chain-ids for EVM chains are numbers!
+        fn chain_type(&self) -> Option<ChainType> {
+            let chains = map_macro::hash_map! {
```

Comment:
I don't like that we spawn this hashmap on each `chain_type()` call. In theory we shouldn't have to call it more than once, but would be nice to ensure it's not generated more than once

Reply:
Ok, something with OnceCell or something. Because static variables can't really be used here, because we are using constants arrays

Reply:
> Ok, something with OnceCell or something. Because static variables can't really be used here, because we are using constants arrays

Also maybe this https://crates.io/crates/phf, haven't tried, but look like would fit here

Reply:
I have already tried

Reply:
> I have already tried and it's not working because they don't support this pattern (ARCHWY[0]) for instance

Ah, that's unfortunate